src/charm.py

```diff
@@ -96,12 +95,54 @@ def _are_relations_ready(self) -> bool:
         return self._tls.get_tls_relation() is not None
 
     def _reconcile(self) -> None:
-        """Reconcile charm status based on configuration and integrations."""
+        """Reconcile charm status based on configuration and integrations.
+
+        Raises:
+            RuntimeError: when initializing the lightkube client fails,
+            or when creating the gateway resource fails.
+        """
+        try:
+            # Set field_manager for server-side apply when patching resources
+            # Keep this consistent across client initializations
+            kubeconfig = KubeConfig.from_service_account()
+            client = Client(config=kubeconfig, field_manager=self.app.name)
+        except ConfigError as exc:
+            LOGGER.error("Error initializing the lightkube client: %s", exc)
+            raise RuntimeError("Error initializing the lightkube client.") from exc
+
+        try:
+            gateway_resource_definition = GatewayResourceDefinition.from_charm(self)
+        except InvalidCharmConfigError as exc:
+            LOGGER.error("Invalid charm config: %s", exc.msg)
+            self.unit.status = BlockedStatus("Invalid charm configuration")
+            return
+
         tls_certificates_relation = self._tls.get_tls_relation()
         if not tls_certificates_relation:
             self.unit.status = BlockedStatus("Waiting for TLS.")
             return
-        self.unit.status = ActiveStatus()
+
+        gateway_resource_manager = GatewayResourceManager(
+            namespace=gateway_resource_definition.namespace,
+            labels=self._labels,
+            client=client,
+        )
+
+        try:
+            gateway = gateway_resource_manager.define_resource(gateway_resource_definition)
+        except (CreateGatewayError, InvalidResourceError) as exc:
+            LOGGER.error("Error creating the gateway resource %s", exc)
+            raise RuntimeError("Cannot create gateway.") from exc
+        except KuberentesCreateResourceError as exc:
+            self.unit.status = BlockedStatus(exc.msg)
+            return
+
+        self.unit.status = WaitingStatus("Waiting for gateway address")
+        if gateway_address := gateway_resource_manager.gateway_address(gateway.metadata.name):
+            self.unit.status = ActiveStatus(f"Gateway addresses: {gateway_address}")
+        else:
+            self.unit.status = WaitingStatus("Gateway address unavailable")
```

Comment:
Is there any action that should be taken here?

Reply:
If the gateway address can't be fetched then that means k8s didn't assign the gateway and IP address / endpoint. Unfortunately since the charm doesn't know how k8s is setup, or even what the gateway controller is, it can't offer much help here.

Reply:
Hm...but if its waitingstatus, could be changed/solved by some other event? Or if not, should be an error?

Reply:
Either way works imho, the charm doesn't necessarily need to go into Error state as it's still functional (the gateway resource was created successfully in this case and tls service is working fine too). On the other hand if the k8s cluster is misconfigured in someway then an operator can reconfigure the cluster so that a LB ip address is assigned to the charm, in which case we'd want a blocked status. But the k8s cluster can also be taking some time to assign the IP address, in which case there's nothing to be done aside from wait for the cluster to do its job. 

All in all I think a waiting status is a good middle ground here, what do you think?

Reply:
Good idea, I'll add it in a follow-up PR.